File: hydra-haskell/src/main/haskell/Hydra/Lib/Equality.hs
-- | Haskell implementations of hydra.lib.equality primitives. These simply make use of derived Eq.

module Hydra.Lib.Equality where

import Hydra.Util

import Data.Int


-- | Compare two values and return a Comparison.
compare :: Ord a => a -> a -> Comparison
compare x y
  | x < y     = ComparisonLessThan
  | x > y     = ComparisonGreaterThan
  | otherwise = ComparisonEqualTo

-- | Check if two values are equal.
equal :: Eq a => a -> a -> Bool
equal = (==)

-- | Check if first value is greater than second.
gt :: Ord a => a -> a -> Bool
gt = (>)

-- | Check if first value is greater than or equal to second.
gte :: Ord a => a -> a -> Bool
gte = (>=)

-- | Return the identity of a value.
identity :: a -> a
identity = id

-- | Check if first value is less than second.
lt :: Ord a => a -> a -> Bool
lt = (<)

-- | Check if first value is less than or equal to second.
lte :: Ord a => a -> a -> Bool
lte = (<=)

-- | Return the maximum of two values.
max :: Ord a => a -> a -> a
max = Prelude.max

-- | Return the minimum of two values.
min :: Ord a => a -> a -> a
min = Prelude.min

File: hydra-python/src/main/python/hydra/lib/equality.py
"""Python implementations of hydra.lib.equality primitives."""

from __future__ import annotations
from typing import Any, TypeVar
import hydra.util

A = TypeVar('A')


def compare(x: Any, y: Any) -> hydra.util.Comparison:
    """Compare two values and return a Comparison."""
    if x < y:
        return hydra.util.Comparison.LESS_THAN
    elif x > y:
        return hydra.util.Comparison.GREATER_THAN
    else:
        return hydra.util.Comparison.EQUAL_TO


def equal(a: Any, b: Any) -> bool:
    """Check if two values are equal."""
    return a == b


def gt(a: Any, b: Any) -> bool:
    """Check if first value is greater than second."""
    return a > b


def gte(a: Any, b: Any) -> bool:
    """Check if first value is greater than or equal to second."""
    return a >= b


def identity(a: A) -> A:
    """Return the identity of a value."""
    return a


def lt(a: Any, b: Any) -> bool:
    """Check if first value is less than second."""
    return a < b


def lte(a: Any, b: Any) -> bool:
    """Check if first value is less than or equal to second."""
    return a <= b


def max(a: Any, b: Any) -> Any:
    """Return the maximum of two values."""
    return a if a > b else b


def min(a: Any, b: Any) -> Any:
    """Return the minimum of two values."""
    return a if a < b else b
